pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go

```
@@ -335,30 +308,108 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
+// exists yet to allow new rollout superceded by the new config version.
+func (c *DeploymentConfigController) cancelRunningRollouts(config *deployapi.DeploymentConfig, existingDeployments []*v1.ReplicationController, cm *RCControllerRefManager) error {
+	awaitingCancellations := false
+	for i := range existingDeployments {
+		deployment := existingDeployments[i]
+		// Skip deployments with an outcome.
+		if deployutil.IsTerminatedDeployment(deployment) {
+			continue
+		}
+		// Cancel running deployments.
+		awaitingCancellations = true
+		if deployutil.IsDeploymentCancelled(deployment) {
+			continue
+		}
+
+		// Retry faster on conflicts
+		var updatedDeployment *v1.ReplicationController
+		err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			rc, err := c.rcLister.ReplicationControllers(deployment.Namespace).Get(deployment.Name)
+			if kapierrors.IsNotFound(err) {
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+			// We need to make sure we own that RC or adopt it if possible
+			isOurs, err := cm.ClaimReplicationController(rc)
+			if err != nil {
+				return fmt.Errorf("error while deploymentConfigController claiming the replication controller %s/%s: %v", rc.Namespace, rc.Name, err)
+			}
+			if !isOurs {
+				return nil
+			}
+
+			copied, err := deployutil.DeploymentDeepCopyV1(rc)
+			if err != nil {
+				return err
+			}
+			copied.Annotations[deployapi.DeploymentCancelledAnnotation] = deployapi.DeploymentCancelledAnnotationValue
+			copied.Annotations[deployapi.DeploymentStatusReasonAnnotation] = deployapi.DeploymentCancelledNewerDeploymentExists
+			updatedDeployment, err = c.rn.ReplicationControllers(copied.Namespace).Update(copied)
+			return err
+		})
+		if err != nil {
+			c.recorder.Eventf(config, v1.EventTypeWarning, "DeploymentCancellationFailed", "Failed to cancel deployment %q superceded by version %d: %s", deployment.Name, config.Status.LatestVersion, err)
+			return nil
```

We should return this error, otherwise bad things may happen, right?